### PR TITLE
KAFKA-3488: Avoid failing of unsent requests in consumer where possible

### PR DESCRIPTION
**Pull request by rajinisivaram**
Fail unsent requests only when returning from KafkaConsumer.poll().

### CODE REVIEW

**Comment by rajinisivaram** [comment]
@hachikuji Can you take a look at my initial attempt to reduce unsent requests? This is the simplest solution I could come up with, but I am not sure if I have missed something. Rather than clearing unsent requests for every `ConsumerNetworkClient.poll`, the PR fails unsent requests only when returning from `KafkaConsumer.poll` after the loop of `ConsumerNetworkClient.poll`s. It fixes the problem with failures during handshake, but I wasn't sure if there was a reason why unsent requests had to be cleared on every iteration.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java [hunk 1]
```diff
@@ -228,7 +229,8 @@ private void poll(long timeout, long now, boolean executeDelayedTasks) {
         trySend(now);
 
         // fail all requests that couldn't be sent
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java]: The comment should either be removed, moved inside the `if` or updated. Personally, it doesn't seem to add much over the actual code, so I would remove it. But let's see what @hachikuji says.

**Comment by hachikuji** [comment]
@rajinisivaram At a glance, the approach seems interesting. I'll look more carefully this afternoon.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 2]
```diff
@@ -878,6 +878,7 @@ public void assign(Collection<TopicPartition> partitions) {
                 remaining = timeout - elapsed;
             } while (remaining > 0);
 
+            client.quickPoll(false, false);
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Can you explain why this case is different than the one above? Why do we not fail unsent requests?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java [hunk 3]
```diff
@@ -228,7 +229,8 @@ private void poll(long timeout, long now, boolean executeDelayedTasks) {
...
-        failUnsentRequests();
+        if (failUnsent)
+            failUnsentRequests();
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java]: Rather than trying to control when and when not to fail unsent requests, I wonder if we should use "request.timeout.ms" to selectively fail only those requests whose timeouts have expired? Then we could remove the `failUnsent` flag everywhere and just call something like `timeoutExpiredRequests()`. In that case, we could remove SendFailedException entirely and use TimeoutException. What do you think?

The only problem is that I'm not sure whether this could cause strange behavior on coordinator fail-over. For example, suppose that we try to send a commit just before the coordinator crashes. In that case, the commit could get stuck in the request queue. We may find the new coordinator and the old coordinator may even come back before the commit's timeout has expired. In that case, if the commit eventually is sent to the old coordinator, then we'd suddenly see a NOT_COORDINATOR_FOR_GROUP error, which would cause us to find the coordinator again unnecessarily. Maybe all we need to address edge cases like this is a way to fail any pending requests to the coordinator when we find that it has failed.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 2 above)
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Oops, sorry, that was a mistake. I had added that line specifically to fail unsent requests :-)

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java (hunk 3 above)
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java]: @hachikuji Thank you for the review. I was not sure of the implications of not failing unsent requests. Hence tried to keep the code as close to the existing semantics as possible. I like the idea of only expiring requests on timeout. Are coordinator changes always associated with a connection failure? If they are, then it shouldn't be hard, otherwise it may be harder to ensure all the edge cases are handled.

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java]: Coordinator changes are _usually_ associated with disconnects, but not necessarily. We have a `coordinatorDead()` function in `AbstractCoordinator` and I was thinking we could make a call to `client.failPendingSends(coordinator)` or something like that at the same time that we set the coordinator to null. I think that would make the behavior the same as the existing code. There may be in-flight sends to the coordinator at the time that we call `coordinatorDead()`, but that is true currently as well, and at least this would prevent any new requests from being sent.

Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java]: @hachikuji Thank you for the explanation. I will update the PR.

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java]: Sounds good. I'm also ok with a more incremental change if it ends up being more complex than I suggested.

**Comment by rajinisivaram** [comment]
@hachikuji I have updated the PR with the fix you suggested to fail requests only on expiry. I used session timeout rather than request timeout as expiry, but will be happy to change it to request timeout if that is more appropriate. Can you take a look? Thank you.

**Comment by rajinisivaram** [comment]
On second thought, request timeout looks better as expiry timeout since session timeout can be quite large. Have updated the PR.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 4]
```diff
@@ -521,6 +522,7 @@ public boolean coordinatorUnknown() {
     protected void coordinatorDead() {
         if (this.coordinator != null) {
             log.info("Marking the coordinator {} dead for group {}", this.coordinator, groupId);
+            client.failUnsentRequests(this.coordinator, GroupCoordinatorNotAvailableException.INSTANCE);
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:525]: I think this exception passes through to the OffsetCommitCallback in `ConsumerCoordinator.commitOffsetsAsync()`. I wonder if we should wrap it in a CommitFailedException in that case since it doesn't give much context to the user. This seems orthogonal to this patch, so we can probably do it separately.

File: core/src/main/scala/kafka/admin/AdminClient.scala [hunk 5]
```diff
@@ -43,21 +43,15 @@ class AdminClient(val time: Time,
...
                    api: ApiKeys,
                    request: AbstractRequest): Struct = {
-    var now = time.milliseconds()
-    val deadline = now + requestTimeoutMs
     var future: RequestFuture[ClientResponse] = null
 
-    do {
-      future = client.send(target, api, request)
-      client.poll(future)
+    future = client.send(target, api, request)
```
Comment (hachikuji) [inline core/src/main/scala/kafka/admin/AdminClient.scala:48]: It has to be a good sign when the usage becomes so much simpler!

Comment (ijuma) [inline core/src/main/scala/kafka/admin/AdminClient.scala:48]: Since we're changing this anyway, can we please fix the issue in this PR https://github.com/apache/kafka/pull/946 ?

Comment (hachikuji) [inline core/src/main/scala/kafka/admin/AdminClient.scala:48]: Does it not fix it? Are there still cases where we block longer than the request timeout?

Comment (ijuma) [inline core/src/main/scala/kafka/admin/AdminClient.scala:48]: I hoped it did, but I didn't check myself. :)

**Comment by hachikuji** [comment]
@rajinisivaram LGTM. I ran all the client system tests to be sure there weren't any obvious problems. One minor note: there's a TODO at the top of ConsumerNetworkClient that I think we can get rid of now. 

I'll ask @ewencp if he has time to review since he's also worked in this code quite a bit.

**Comment by rajinisivaram** [comment]
@hachikuji Thank you for the review and for running the system tests. I have removed the TODO in ConsumerNetworkClient.

**Comment by ewencp** [comment]
LGTM, thanks @rajinisivaram!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 4 above)
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:525]: Is there a separate JIRA filed for this case?

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:525]: Added it just now: https://issues.apache.org/jira/browse/KAFKA-3527
